### PR TITLE
DPDK backend: Fixing criteria for table match key consolidation into metadata

### PR DESCRIPTION
**Pull request by usha1830**
The criteria for making copy of header match keys in metadata was being incorrectly checked by the below condition:
`if (key->expression->toString() != firstKey->toString())`

Updated  to check various conditions in which the copy should be done. 
Added a few tests for covering various such scenarios.

### CODE REVIEW

**Comment by usha1830** [comment]
@mbudiu-vmw @hanw Could you please review this PR?

File: backends/dpdk/dpdkArch.cpp [hunk 1]
```diff
@@ -1006,30 +1006,44 @@ const IR::Node* CopyMatchKeysToSingleStruct::preorder(IR::Key* keys) {
...
         return keys;
     }
 
-    firstKey = keys->keyElements.at(0)->expression;
+    cstring firstKeyStr = "";
+    bool firstKeyHdr = false;
+
+    if (auto firstKeyField = keys->keyElements.at(0)->expression->to<IR::Member>()) {
+        firstKeyStr = firstKeyField->expr->toString();
+        if (firstKeyStr.startsWith("h"))
```
Comment (mbudiu-vmw) [inline backends/dpdk/dpdkArch.cpp:1020]: can you explain what 'h' means?

Comment (usha1830) [reply backends/dpdk/dpdkArch.cpp:1020]: A previous pass converts all header fields to the form "h.<header_name>.<field_name> and similarly metadata fields are prefixed with "m.". So, here we are checking if the match key is part of a header by checking the "h" prefix.

Comment (mbudiu-vmw) [reply backends/dpdk/dpdkArch.cpp:1020]: That must be a comment in the code.

Comment (usha1830) [reply backends/dpdk/dpdkArch.cpp:1021]: Added a comment for the same.

Comment (mbudiu-vmw) [reply backends/dpdk/dpdkArch.cpp:1021]: I liked your comment on github better than the one in the code.

**Review by mbudiu-vmw — APPROVED** [review]
I will approve the implementation.
You should use rebase to merge the code from the repository, as described in the README.
You have picked up many unrelated commits which will be very confusing.
Maybe you can actually squash all these commits resubmit the PR with just one commit?

**Comment by usha1830** [comment]
> Maybe you can actually squash all these commits resubmit the PR with just one commit?
I have created a new PR with just one commit https://github.com/p4lang/p4c/pull/2867

**Comment by usha1830** [comment]
Created a New PR with these changes. #2867
Closing this PR.

**Pull request closed without merging**